File: Leetcode/2438 - Range Product Queries of Powers/solution.js
var productQueries = function(n, queries) {
    const MOD = 1e9 + 7;
    const powers = [];

    for (let p = 1; p <= n; p <<= 1) {
        if ((n & p) !== 0) {
            powers.push(p);
        }
    }

    const answers = [];
    for (const [left, right] of queries) {
        let product = 1;
        for (let i = left; i <= right; i++) {
            product = (product * powers[i]) % MOD;
        }
        answers.push(product);
    }

    return answers;
};

File: Leetcode/2438 - Range Product Queries of Powers/solution.cpp
vector<int> productQueries(int n, vector<vector<int>>& queries) {
    const int MOD = 1e9 + 7;
    vector<int> powers;

    for (int p = 1; p <= n; p <<= 1) {
        if ((n & p) != 0) {
            powers.push_back(p);
        }
    }

    vector<int> answers;
    for (auto &q : queries) {
        int left = q[0], right = q[1];
        long long product = 1;
        for (int i = left; i <= right; i++) {
            product = (product * powers[i]) % MOD;
        }
        answers.push_back((int)product);
    }

    return answers;
}